lib/index.js

```diff
@@ -55,82 +55,79 @@ function setupRateLimitEventExt(server, options) {
   const extractKey = options.extractKey;
   const onError = options.onError;
 
-  const extractKeyAndTakeToken = function(limitd, request, reply, type) {
-    extractKey(request, reply, (err, key) =>{
-      if (err) { return reply(err); }
-
-      if (!limitd) {
-        // limitd is not connected, do not fail!
-        return reply.continue();
-      }
-
-      limitd.take(type, key, (err, currentLimitResponse) => {
-        if (err){
-          if (onError) { return onError(err, reply); }
-          // by default we don't fail if limitd is unavailable
-          return reply.continue();
-        }
-
-        const oldMinimumLimitResponse = request.plugins.patova && request.plugins.patova.limit
-        const newMinimumLimitResponse = getMinimumLimit(currentLimitResponse, oldMinimumLimitResponse)
-
-        request.plugins.patova = request.plugins.patova || {};
-        request.plugins.patova.limit = newMinimumLimitResponse;
-
-        if (newMinimumLimitResponse.conformant) {
-          // We continue only if the request is conformat so far
-          return reply.continue();
-        }
-
-        const error = Boom.tooManyRequests();
-        error.output.headers = new RateLimitHeaders(
-          newMinimumLimitResponse.limit,
-          newMinimumLimitResponse.remaining,
-          newMinimumLimitResponse.reset);
-
-        reply(error);
-      });
-    });
+  const extractKeyAndTakeToken = async function(limitd, request, handler, type) {
+    const key = extractKey(request, handler);
+    if (err) { return err; }
+
+    if (!limitd) {
+      // limitd is not connected, do not fail!
+      return handler.continue;
+    }
+
+    try {
+      const currentLimitResponse = await limitd.take(type, key);
+
+    } catch(err) {
+      if (onError) { return onError(err, reply); }
+      // by default we don't fail if limitd is unavailable
+      return handler.continue;
+    }
+
+
+    const oldMinimumLimitResponse = request.plugins.patova && request.plugins.patova.limit
+    const newMinimumLimitResponse = getMinimumLimit(currentLimitResponse, oldMinimumLimitResponse)
+
+    request.plugins.patova = request.plugins.patova || {};
+    request.plugins.patova.limit = newMinimumLimitResponse;
+
+    if (newMinimumLimitResponse.conformant) {
+      // We continue only if the request is conformat so far
+      return handler.continue;
+    }
+
+    const error = Boom.tooManyRequests();
+    error.output.headers = new RateLimitHeaders(
+      newMinimumLimitResponse.limit,
+      newMinimumLimitResponse.remaining,
+      newMinimumLimitResponse.reset);
+
+    throw error;
   };
 
-  const getType = function(request, reply, callback) {
+  const getType = async function(request, handler, callback) {
     const type = options.type;
 
     if (typeof type !== 'function') {
-      return process.nextTick(() => callback(null, type));
+      return process.nextTick(() => Promise.resolve(type));
```

Comment:
I think `process.nextTick` does not return a `Promise` (or is it promisified somehow?).